spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java

```diff
@@ -119,12 +120,24 @@ private void addInstancesToList(List<ServiceInstance> instances, String serviceI
 	public List<String> getServices() {
 		String aclToken = properties.getAclToken();
 
+		Map<String, List<String>> services;
 		if (StringUtils.hasText(aclToken)) {
-			return new ArrayList<>(client.getCatalogServices(QueryParams.DEFAULT, aclToken).getValue()
-					.keySet());
+			services = client.getCatalogServices(QueryParams.DEFAULT, aclToken).getValue();
```

Comment:
These methods should use the `getCatalogServices()` method that takes a tag parameter (which can be null, which is the default value for `properties.defaultQueryTag`). Otherwise, for a large deployment this would be very inefficient.

Reply:
Thanks for taking a look at this PR. I don't see a `getCatalogService()` method that can be used for this. The only ones which take a tag also take a service. The problem this is trying to solve is to query all services with a given tag, so the service name is not known upfront.

Reply:
I see them. Maybe you need to rebase from master to get the latest version of consul-api.

Reply:
I see 2 methods which accept the `tag` parameter, and both require a `serviceName` to be provided
```
public Response<List<com.ecwid.consul.v1.catalog.model.CatalogService>> getCatalogService(String serviceName, String tag, QueryParams queryParams)
public Response<List<com.ecwid.consul.v1.catalog.model.CatalogService>> getCatalogService(String serviceName, String tag, QueryParams queryParams, String token)
```
I've looked at the [implementation](https://github.com/Ecwid/consul-api/blob/master/src/main/java/com/ecwid/consul/v1/catalog/CatalogConsulClient.java#L141) and `getCatalogServices(QueryParams queryParams)` used here makes a call to Consul's [list services' endpoint](https://www.consul.io/api/catalog.html#list-services) which does *not* accept a list of tags. The list services endpoint returns the entire catalog along with tags in a single request regardless of how large the deployment is.

The methods that accept a tag for filtering calls Consul's [list node for services' endpoint](https://www.consul.io/api/catalog.html#list-nodes-for-service) which does take in a tag but requires a specific service to be provided. [(implementation)](https://github.com/Ecwid/consul-api/blob/master/src/main/java/com/ecwid/consul/v1/catalog/CatalogConsulClient.java#L183)

Reply:
Thanks, you are right. I guess our current implementation is no less performant :-)

Reply:
Great, I'm relieved to hear that. I've only started working with Consul recently and was worried I missed something.